File: python/backend-urllib/05-subdivision-follow-related.py
"""Subdivision validation with follow_related example using urllib.request"""

import urllib.request
import urllib.parse
import json
import sys

# Set your CountriesDB private API key here
API_KEY = 'your_private_api_key_here'
API_BASE = 'https://api.countriesdb.com'

if not API_KEY or API_KEY == 'your_private_api_key_here':
    raise SystemExit('Please set your API_KEY in the script')

# 5. Subdivision validation with follow_related
def main():
    try:
        print('Validating subdivision code: FR-40 for country: FR (with follow_related)')
        url = f'{API_BASE}/api/validate/subdivision'
        
        data = json.dumps({
            'country': 'FR',
            'code': 'FR-40',
            'follow_related': True
        }).encode('utf-8')
        
        req = urllib.request.Request(
            url,
            data=data,
            headers={
                'Authorization': f'Bearer {API_KEY}',
                'Content-Type': 'application/json',
            }
        )
        
        with urllib.request.urlopen(req) as response:
            result = json.loads(response.read().decode('utf-8'))
            print('Response:', json.dumps(result, indent=2))
            
            if not result.get('valid'):
                raise ValueError('Subdivision validation failed')
            
            print('Subdivision is valid')
    except urllib.error.HTTPError as exc:
        error_body = exc.read().decode('utf-8')
        print(f'Request failed: {exc.code} {exc.reason}', file=sys.stderr)
        print(f'Response: {error_body}', file=sys.stderr)
        sys.exit(1)
    except Exception as exc:
        print(f'Validation failed: {exc}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: ruby/backend-nethttp/05-subdivision-follow-related.rb
require 'bundler/setup'
require 'net/http'
require 'json'
require 'uri'

# Set your CountriesDB private API key here
api_key = 'YOUR_API_KEY' # Replace with your actual API key
if api_key == 'YOUR_API_KEY'
  abort 'Please set your API_KEY in the script'
end

def call(api_key, path, payload)
  uri = URI("https://api.countriesdb.com#{path}")
  
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = true
  
  request = Net::HTTP::Post.new(uri.path)
  request['Authorization'] = "Bearer #{api_key}"
  request['Content-Type'] = 'application/json'
  request.body = payload.to_json
  
  response = http.request(request)
  body = JSON.parse(response.body)
  raise "Request failed: #{response.code} #{body}" unless response.is_a?(Net::HTTPSuccess)
  
  body
end

# 5. Subdivision validation with follow_related
begin
  puts 'Validating subdivision code: FR-40 (country: FR, with follow_related)'
  result = call(api_key, '/api/validate/subdivision', {
    country: 'FR',
    code: 'FR-40',
    follow_related: true
  })
  puts "Response: #{JSON.pretty_generate(result)}"
  
  unless result['valid']
    raise 'Subdivision validation failed'
  end
  
  puts '✅ Subdivision is valid'
rescue => e
  puts "❌ Validation failed: #{e.message}"
  exit 1
end

